File: Merge-Sort/mergeSort.cpp
#include <iostream>
#include <limits>
#include "mergeSort.h"

void printArray(int a[], int len){
	for(int i=1; i<len; i++){
		std::cout << a[i-1] << ", ";
	}
	std::cout << a[len-1] << std::endl;
}


int main(int argc, char** argv){

	int numArray[] = {5, 8, 3, 5, 2, 7, 6, 1};
	int len = sizeof(numArray)/sizeof(numArray[0]);


	std::cout << "Original array: ";
	printArray(numArray, len);

	mergeSort(numArray, 0, len-1);


	std::cout << "Sorted array: ";
	printArray(numArray, len);
	
	return 0;
}

File: Merge-Sort/mergeSort.py
#!/usr/bin/env python3

def mergeSortJoin(numArray, start, mid, end):

	leftArray = []
	rightArray = []

	for element in numArray[start:mid+1]:
		leftArray.append(element)

	for element in numArray[mid+1:end+1]:
		rightArray.append(element)

	leftIndex = 0
	rightIndex = 0

	for i in range(start,end+1):
		if leftIndex >= len(leftArray):
			numArray[i] = rightArray[rightIndex]
			rightIndex+=1
			continue

		if rightIndex >= len(rightArray):
			numArray[i] = leftArray[leftIndex]
			leftIndex +=1
			continue

		if(leftArray[leftIndex] < rightArray[rightIndex]):
			numArray[i] = leftArray[leftIndex]
			leftIndex +=1
		else:
			numArray[i] = rightArray[rightIndex]
			rightIndex +=1


def mergeSort(numArray, start, end):
	if start >= end:
		return

	mid = (start+end)//2

	mergeSort(numArray, start, mid)
	mergeSort(numArray, mid+1, end)
	mergeSortJoin(numArray, start, mid, end)


if __name__ == "__main__":
	numArray = [5, 8, 3, 5, 2, 7, 6, 1]
	arrayLen = len(numArray)

	print("Original Array: ", numArray)
	mergeSort(numArray, 0, arrayLen-1)
	print("Sorted Array: ", numArray)